Classifier results for different genes

brankaj opened the pull request:
Hi all,

Since MVP is going to be released soon and open to the public, maybe it would be useful to provide some kind of demo to explain how the package works. It would be like providing tutorial/examples for the first-time users. 
Also, since most likely sklearn.linear_model.SGDClassifier is going to be used as a classifier, it would be nice to see the results for different genes, not only TP53. During the last meeting there was a discussion regarding this. Since it would take really long time to run the code for all genes, maybe it's better to focus on the group of genes that are considered dominant by a certain criteria. 
I have submitted pull requests for two files. Both files output AUROCs for a list of genes. First file is a code that determines AUROCs for a list of tumor suppressor genes. The list is not exhaustive and it is copied from https://www.cancerquest.org/cancer-biology/cancer-genes/tumor-suppressors/tumor-suppressor-table. The second code runs the classifier for the genes that are associated with the lung cancer. This list of genes is obtained from https://ghr.nlm.nih.gov/condition/lung-cancer#genes.

Not sure is this going to be useful to the researchers in the field, but I thought giving it a try. 
All feedback welcome!

brankaj committed:
Classifier results for different genes

Hi all,
Since MVP is going to be released soon and open to the public, maybe it would be useful to provide some kind of demo to explain how the package works. It would be like providing tutorial/examples for the first-time users. 
Also, since most likely sklearn.linear_model.SGDClassifier is going to be used as a classifier, it would be nice to see the results for different genes, not only TP53. During the last meeting there was a discussion regarding this. Since it would take really long time to run the code for all genes, maybe it's better to focus on the group of genes that are considered dominant by a certain criteria. 
I have submitted pull requests for two files. Both files output AUROCs for a list of genes. First file is a code that determines AUROCs for a list of tumor suppressor genes. The list is not exhaustive and it is copied from https://www.cancerquest.org/cancer-biology/cancer-genes/tumor-suppressors/tumor-suppressor-table. The second code runs the cla

dhimmel commented on algorithms/2.TCGA-MLexample-LungCancer.py:
```diff
@@ -0,0 +1,166 @@
+
+# GENES: AKT1, ALK, BRAF, DDR2, EGFR, ERBB2, KRAS, MAP2K1, NRAS, PIK3CA, PTEN, RET, RIT1, ROS1
+
+list_of_tsgenes = [207,238,673,4921,1956,2064,3845,5604,4893,5290,5728,5979,6016,6098]
```
Consider using a dictionary of entrez_gene_id to symbol. This way you can easily make the plot labels use gene symbols which are more recognizable.

brankaj replied:
Thank you for the feedback! I used ids listed on the NCBI website. I made sure that the numbers correspond to the Homo sapiens column :bowtie: 
I will prepare one notebook with the incorporated changes. Time to learn about dictionaries. 
Btw, since I do not see the option to move the files to the explore dictionary, is it easier that I close/delete this pull request and start a new one with one notebook?

dhimmel replied:
> Thank you for the feedback! I used ids listed on the NCBI website. I made sure that the numbers correspond to the Homo sapiens column

That's the correct way to do it. See [this example](https://github.com/cognoma/cancer-data/blob/b054721bbde1ed0abc19f395eb7629728870cb7c/scripts/3.explore-mutations.py#L100-L109) for what I mean by the dictionary. The goal is to make your plot labels show gene symbols rather than entrez GeneIDs.

dhimmel commented on algorithms/2.TCGA-MLexample-LungCancer.ipynb:
```diff
@@ -0,0 +1,311 @@
+    "        feature_select,  # Feature selection\n",
+    "        StandardScaler(),  # Feature scaling\n",
+    "        clf_grid)\n",
+    "    %%time\n",
```
I think this `%%time` should be removed -- it doesn't look like it's really working :smile_cat:

dhimmel reviewed (changes requested):
Really nice exploration! It was interesting to see how overfitting got worse when the positive prevalence decreased.

Let's move this to the `explore` directory rather than the `algorithms` where it is now.

Also, would it be possible to consolidate this analysis to one notebook. DRY -- do not repeat yourself. Right now almost all the code is repeated. The one notebook can create two plots: a lung cancer gene plot and a suppressor gene plot but most of the code will be shared.

dhimmel commented:
> Btw, since I do not see the option to move the files to the explore ~~dictionary~~ directory, is it easier that I close/delete this pull request and start a new one with one notebook?

No need to create a new pull request. You can just make the changes to your `brankaj-patch-2` branch. [Here's some info](https://githowto.com/moving_files) on how to move files that are tracked in git.

brankaj committed:
Delete 2.TCGA-MLexample-LungCancer.ipynb

brankaj committed:
Delete 2.TCGA-MLexample-LungCancer.py

brankaj committed:
Delete 2.TCGA-MLexample-TumorSuppressors.ipynb

brankaj committed:
Delete 2.TCGA-MLexample-TumorSuppressors.py

brankaj committed:
Merge pull request #2 from cognoma/master

Updating fork

brankaj committed:
Merge pull request #3 from brankaj/master

Update branch

brankaj committed:
Adding one notebook with two plots

Two plots are generated - a lung cancer gene plot and a suppressor gene plot

brankaj committed:
Adding .py file

brankaj committed:
Delete .py file

brankaj committed:
Delete .ipynb file

brankaj committed:
Adding one notebook with two plots

dhimmel commented on explore/Classifier_results-different_genes.py:
```diff
@@ -0,0 +1,246 @@
+    # Set aside 10% of the data for testing, this can only be done where the number of mutations is large enough
+    X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.1, random_state=0)
+    'Size: {:,} features, {:,} training samples, {:,} testing samples'.format(len(X.columns), len(X_train), len(X_test))
+    def fs_mad(x, y):
```
You can define `fs_mad` and `get_threshold_metrics` outside of the for loop. Currently, every iteration of the loop redefines the function which is unnecessary because the function does not change.

dhimmel commented on explore/Classifier_results-different_genes.py:
```diff
@@ -0,0 +1,246 @@
+    metrics_train = get_threshold_metrics(y_train, y_pred_train)
+    metrics_test = get_threshold_metrics(y_test, y_pred_test)
+    
+    AUROCS_train += [metrics_train['auroc']]
```
You can use `append` to add a single element to the end of a list:

``` python
AUROCS_train.append(metrics_train['auroc'])
```

dhimmel reviewed (changes requested):
Nice analysis. It's encouraging to see that all classifiers had testing performance with AUROC > 50%.

Just two minor comments.

brankaj committed:
Delete Classifier_results-different_genes.ipynb

brankaj committed:
Delete Classifier_results-different_genes.py

brankaj committed:
Notebook with implemented two changes

Definition of functions outside the loop
Append command instead of +=

gwaybio commented on explore/Classifier_results-different_genes.py:
```diff
@@ -0,0 +1,252 @@
+for GENE in list_of_genes:
+    
+    y = Y[GENE]
+    # Here are the percentage of tumors with NF1
```
Comment mentioning `NF1` no longer applies - this could be confusing inside the loop over genes

gwaybio commented on explore/Classifier_results-different_genes.py:
```diff
@@ -0,0 +1,252 @@
+
+# In[9]:
+
+AUROCS_train_lung = AUROCS_train[0:len(list(genes_LungCancer.values()))]
```
no need to convert `genes_LungCancer` to a list, this should work fine:

``` python
AUROCS_train_lung = AUROCS_train[0:len(genes_LungCancer)]
```

This change would propagate through lines below as well

gwaybio reviewed (commented):
This is a fantastic analysis - It is really nice to show that Cognoma works well for a variety of genes! I really like the bar charts as well...

This may be beyond the scope of the current pull request, but plotting performance for some of the oncogenes would be helpful as well (However, I do see a couple in the lung cancer genes analysis).

brankaj committed:
Delete Classifier_results-different_genes.ipynb

brankaj committed:
Delete Classifier_results-different_genes.py

brankaj committed:
Results for different set of genes

This file contains results for:
-Lung Cancer genes
-Tumor suppressors
-Oncogenes

Oncogenes are chosen based on the categories listed on https://en.wikipedia.org/wiki/Oncogene#Classification.
The required changes are incorporated as well as two other things:
1. The get_threshold_metrics() function is removed since the code only plots AUROCs. The AUROC value is directly computed using roc_auc_score().
2. The suppression of  joblib warning is added at the end of the for loop code. For some reason, I get a warning regarding the 'the shape of non-C contiguous array'.

brankaj commented:
Thanks a lot for the comments. The details regarding the latest file are given in the git commit message. Let me know if you prefer that I restore the get_threshold_metrics() function.

dhimmel commented on explore/Classifier_results-different_genes.py:
```diff
@@ -0,0 +1,291 @@
+    '3791':'KDR', #receptor tyrosine kinases
+    '25':'ABL1', #Cytoplasmic tyrosine kinases
+    '6714':'SRC', #Cytoplasmic tyrosine kinases
+    '5894':'Raf-1',#cytoplasmic serine kinases
```
The [primary symbol](https://www.ncbi.nlm.nih.gov/gene/5894) for this gene is RAF1. In general human gene names are all upercase.

dhimmel commented on explore/Classifier_results-different_genes.py:
```diff
@@ -0,0 +1,291 @@
+    '4771':'NF2',
+    '7157':'TP53', 
+    '5728':'PTEN', 
+    '5925':'Rb',
```
[RB1 is the approved symbol](https://www.ncbi.nlm.nih.gov/gene/5925).

dhimmel reviewed (commented):
Looks great to me. It does seem that a few of your gene names are non-standard. Can you switch to approved symbols? Then we'll be ready to merge!

gwaybio commented on explore/Classifier_results-different_genes.py:
```diff
@@ -0,0 +1,291 @@
+
+# In[12]:
+
+# Plot for oncogenes
```
Fascinating... Generally, the performance is _worse_ for oncogenes - particularly MYC and SRC. MYC would be an interesting case to contrast how well we improve performance by incorporating copy number data (since MYC has copy number gains in several tumor types)

gwaybio replied:
I suspect also that Cognoma will need to be smart about how CNVs are incorporated in the `Y` matrix. My thinking is that for oncogenes a copy number gain would define a 1 whereas for tumor suppressors a copy number loss would define a 1.

brankaj committed:
Delete Classifier_results-different_genes.ipynb

brankaj committed:
Update Classifier_results-different_genes.py

brankaj committed:
Notebook with updated gene symbols

brankaj reviewed (commented):
I am still getting that deprecation warning even with the different placing of the ignore command :/

dhimmel commented:
> I am still getting that deprecation warning even with the different placing of the ignore command :/

Not sure about the warning. You should only have to run the `warnings.ignore` command once, before the code that causes the warning ever gets executed. Have you updated your sklearn to 0.18 (recently done in https://github.com/cognoma/machine-learning/pull/53). This may remove the error.

I'll merge this tomorrow morning, so if you fix the error by then cool, but if not no big deal.

brankaj commented:
Thank you guys for all the comments. I have updated sklearn before running the code, but the warning is still there. I will try to figure out what can be done regarding this issue.
I am also curious about the bad performance for oncogenes. I do not believe it is the classifier 'fault', but I will try different parameter values just to rule out this possibility. Other possibility would be that features or labels for genes like MYC are not that reliable. If this is the case, it is less likely that any classifier can help.

gwaybio commented:
> I do not believe it is the classifier 'fault', but I will try different parameter values just to rule out this possibility. Other possibility would be that features or labels for genes like MYC are not that reliable. If this is the case, it is less likely that any classifier can help.

These are my thoughts at the moment as well. I think we will need to more intelligently define our Y matrix - this is where I think we'll get the most performance gains. This is an active research question right now!
